cpp/augment_773a.cc

```diff
@@ -0,0 +1,167 @@
+#include "DirectoryEntry.h"
+#include "Leader.h"
+#include "MarcUtil.h"
+#include "RegexMatcher.h"
+#include "StringUtil.h"
+#include "Subfields.h"
+#include "util.h"
+
+
+void Usage() {
+    std::cerr << "Usage: " << progname << " [--verbose] marc_input aux_marc_input marc_output\n"
+	      << "       \"marc_input\" is the file that will be augmented and converted.\n"
+	      << "       \"marc_input\" and \"aux_marc_input\" will be scoured for titles that\n"
+	      << "       may be filled into 773$a fields where appropriate.\n"
+              << "       Populates 773$a where it is missing and uplinks exist in 773$x.\n";
+    std::exit(EXIT_FAILURE);
+}
+
+
+static std::unordered_map<std::string, std::string> control_numbers_to_titles_map;
+
+
+bool RecordControlNumberToTitleMapping(MarcUtil::Record * const record, XmlWriter * const /*xml_writer*/,
+		 std::string * const /* err_msg */)
+{
+    ssize_t _245_index;
+    if (likely((_245_index = record->getFieldIndex("245")) != -1)) {
+	const std::vector<std::string> &fields(record->getFields());
+	const Subfields _245_subfields(fields[_245_index]);
+	const std::string title(_245_subfields.getFirstSubfieldValue('a'));
+	if (likely(not title.empty()))
+	    control_numbers_to_titles_map[record->getControlNumber()] = title;
+    }
+
+    return true;
+}
+
+
+void CollectControlNumberToTitleMappings(const bool verbose, const std::vector<File *> &inputs) {
+    std::string err_msg;
+    for (auto &input : inputs) {
+	if (verbose)
+	    std::cout << "Extracting control numbers to title mappings from \"" << input->getPath() << "\".\n";
+	if (not MarcUtil::ProcessRecords(input, RecordControlNumberToTitleMapping, /* xml_writer = */nullptr, &err_msg))
+	    Error("error while looking for control numbers to title mappings: " + err_msg);
+    }
+
+    if (verbose)
+	std::cout << "Found " << control_numbers_to_titles_map.size() << " control number to title mappings.\n";
+}
+
+
+static unsigned patch_count;
+
+
+// Looks for the existence of a 773 field.  Iff such a field exists and 773$a is missing, we try to add it.
+bool PatchUpOne773a(MarcUtil::Record * const record, XmlWriter * const xml_writer, std::string * const /*err_msg*/) {
+    ssize_t _773_index;
+    if ((_773_index = record->getFieldIndex("773")) != -1) {
+	const std::vector<std::string> &fields(record->getFields());
+        const Subfields _773_subfields(fields[_773_index]);
+	if (not _773_subfields.hasSubfield('a') and _773_subfields.hasSubfield('w')) {
+	    const std::string w_subfield(_773_subfields.getFirstSubfieldValue('w'));
+	    if (StringUtil::StartsWith(w_subfield, "(DE-576)")) {
+		const std::string parent_control_number(w_subfield.substr(8));
+		const auto control_number_and_title(control_numbers_to_titles_map.find(parent_control_number));
+		if (control_number_and_title != control_numbers_to_titles_map.end()) {
+		    record->updateField(_773_index, fields[_773_index] + "\x1F""a" + control_number_and_title->second);
```

Comment:
There should be an addSubfield function to hide the ugly separator string (in the future...)